### PR TITLE
[rllib] fixes from dogfooding multi-agent

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

- auto wrap multi-agent dict and tuple spaces by keeping a policy -> preprocessor in the sampler
- add some Q-learning debug stats
- report min, max of custom metrics
- better errors

### CODE REVIEW

**Comment by ericl** [comment]
jenkins retest this please

File: python/ray/rllib/test/test_multi_agent_env.py [hunk 1]
```diff
@@ -64,7 +70,7 @@ def __init__(self, num, increment_obs=False):
         self.last_info = {}
         self.i = 0
         self.num = num
-        self.observation_space = gym.spaces.Discrete(2)
+        self.observation_space = gym.spaces.Discrete(10)
```
Comment (richardliaw) [inline python/ray/rllib/test/test_multi_agent_env.py]: is there a test for the multi-agent preprocessing?

File: python/ray/rllib/test/test_multi_agent_env.py [hunk 2]
```diff
@@ -303,10 +309,20 @@ def testMultiAgentSampleRoundRobin(self):
         # since we round robin introduce agents into the env, some of the env
         # steps don't count as proper transitions
         self.assertEqual(batch.policy_batches["p0"].count, 42)
-        self.assertEqual(batch.policy_batches["p0"]["obs"].tolist()[:10],
-                         [0, 1, 2, 3, 4] * 2)
-        self.assertEqual(batch.policy_batches["p0"]["new_obs"].tolist()[:10],
-                         [1, 2, 3, 4, 5] * 2)
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/test/test_multi_agent_env.py]: 2 was just wrong

File: python/ray/rllib/test/test_multi_agent_env.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/test/test_multi_agent_env.py]: Yeah test_nested_spaces.

**Review by richardliaw — APPROVED** [review]
didn't try out but looks good if tests pass

**Pull request merged**